core/src/main/scala/org/apache/spark/sql/internal/JarUtils.scala

```diff
@@ -0,0 +1,119 @@
+ * This class can not initialize itself after a driver failure. So the callers will have  to make sure that the classloader gets
+ * initialized after a driver startup. Usually it can be achieved by adding classloader at query time.
+ *
+ */
+class ContextJarUtils(sparkContext: SparkContext) extends Logging{
+
+  val workingDir = new File(System.getProperty("user.dir"))
+
+  private val driverJars = new ConcurrentHashMap[String, URLClassLoader]().asScala
+
+  def addDriverJar(key: String, classLoader: URLClassLoader): Option[URLClassLoader] = {
+    driverJars.putIfAbsent(key, classLoader)
+  }
+
+  def addIfNotPresent(key: String, urls: Array[URL], parent: ClassLoader): Unit = {
+    if (driverJars.get(key).isEmpty) {
+      driverJars.putIfAbsent(key, new MutableURLClassLoader(urls, parent))
+    }
+  }
+
+  def removeDriverJar(key: String) = driverJars.remove(key)
+
+
+  def getDriverJar(key: String): Option[URLClassLoader] = driverJars.get(key)
+
+  /**
+   * This method will copy the given jar to Spark root directory and then it will add the same jar to Spark.
+   * This ensures class & jar isolation although it might be repetitive.
+   * @param prefix prefix to be given to the jar name
+   * @param path  original path of the jar
+   */
+  def addToSparkJars(prefix: String, path: String) {
+
+    val callbacks = ToolsCallbackInit.toolsCallback
+    if (callbacks != null) {
+      val localName = path.split("/").last
+      val changedFileName = s"${prefix}-${localName}"
+
+      val changedFile = new File(workingDir, changedFileName)
+      if (!changedFile.exists()) {//After creation removeFromSparkJars() is the only place which can remove this jar
+        println("Adding jar to sc from driver loader" + path)
+        val newFile = callbacks.doFetchFile(path, workingDir, changedFileName)
+        sparkContext.addJar(newFile.getPath)
+        //Setting the local property. Snappy Cluster executors will take appropriate actions
+        sparkContext.
+            setLocalProperty(io.snappydata.Constant.CHANGEABLE_JAR_NAME, newFile.getPath)
+      }
+    }else{
+      sparkContext.addJar(path)
+    }
+  }
+
+  def removeFromSparkJars(prefix: String, path: String): Unit = {
+    def getName(path: String): String = new File(path).getName
+
+    val callbacks = ToolsCallbackInit.toolsCallback
+    if (callbacks != null) {
+      val localName = path.split("/").last
+      val changedFileName = s"${prefix}-${localName}"
+      val jarFile = new File(workingDir, changedFileName)
+
+      if (jarFile.exists()) {
+        jarFile.delete()
+      }
+
+      val keyToRemove = sparkContext.listJars().filter(getName(_) == getName(jarFile.getPath))
+      if (keyToRemove.nonEmpty) {
+        println(s"Removing ${path} from Spark Jars by Driver loader")
+        callbacks.removeAddedJar(sparkContext, keyToRemove.head)
+      }
+      //Remove the jar from all live executors.
+      org.apache.spark.sql.collection.Utils.mapExecutors(
```

Comment:
If one user only wants to change the logic of UDF he will drop and create the function again. If no job other job has been executed between dropFunction and createFunction the new added function will not take effect.